Fix focusResult of Search Editor

mkasenberg opened the pull request:
~The customized value of focusResult was not used if the search editor was opened from the search view. If focusResult is set to false let's ensure that the search editor input is focused instead.~

~Fixed my keybinding:~
```
	{
		"key": "ctrl+shift+f",
		"command": "search.action.openEditor",
		"args": {
			"focusResults": false,
			"resetCursor": false,
			"triggerSearch": true,
		      },
	},
```

Adding the focusResults option to Search Editor options will allow to have a consistent behavior no matter where the triggerSearch has been fired from. If focusResults is set to false, let's ensure that the search editor input is focused.
    
Use cases that used to focus the Search Editor results instead of its input and were not configurable:
- If the Search Editor has been opened from the search view ("Open in editor" button).
~- If a Peek Definition window was open inside the Search Editor results, then at the widget closure, after the next search task has been completed, the focus was moved from the Search Editor input to its results.~

Fixes https://github.com/microsoft/vscode/issues/214369

chandra1109 commented:
"key": "ctrl+shift+f",
		"command": "search.action.openEditor",
		"args": {
			"focusResults": false,
			"resetCursor": false,
			"triggerSearch": true,
		      },
	}, 
Review the following

andreamah commented on src/vs/workbench/contrib/searchEditor/browser/searchEditorActions.ts:
```diff
@@ -173,12 +173,14 @@ export const openNewSearchEditor =
 			editor = await editorService.openEditor(input, { pinned: true }, toSide ? SIDE_GROUP : ACTIVE_GROUP) as SearchEditor;
 		}
 
+		editor.setFocusResults(args.focusResults);
```
I'm not sure how this part helps to fix the bug, since the `args.focusResults` was being passed in previously anyways. Why does this need to be made a class variable?

andreamah replied:
The only change that I see making a difference was the one found here: https://github.com/microsoft/vscode/pull/205914/files#diff-9cc7738b5c4fb266d70547a9cc8b7f6ea60116870c16a677f6658059c768aa45R484-R485, but please correct me if I'm wrong.

mkasenberg replied:
> I'm not sure how this part helps to fix the bug, since the `args.focusResults` was being passed in previously anyways. Why does this need to be made a class variable?

You're right, we might just use casting like this:
```C
const focusResults = (this.priorConfig as LegacySearchEditorArgs).focusResults;
```

> The only change that I see making a difference was the one found here: https://github.com/microsoft/vscode/pull/205914/files#diff-9cc7738b5c4fb266d70547a9cc8b7f6ea60116870c16a677f6658059c768aa45R484-R485

Yes, this is the most important change. All other changes are needed to make the `focusResults` option to work in the use case that was hardcored to `true` here:
https://github.com/microsoft/vscode/blob/313c78eeadbce621c3a5e890b4d1760bb21d7a46/src/vs/workbench/contrib/searchEditor/browser/searchEditorActions.ts#L222

andreamah commented:
Are you talking about opening the search editor from where it says "open in editor" here?
![image](https://github.com/microsoft/vscode/assets/31675041/9eeb0520-cca5-47dd-ad3a-3ca7c88c845e)

The keybindings are solely for running that command from the keyboard shortcut. In your case, that's `ctrl+shift+f`. So your keybinding is saying "when I use `ctrl+shift+f`, run `search.action.openEditor` with the following arguments". This doesn't influence when you open the search editor from other places, like from the search view.

mkasenberg commented:
Sorry, I mixed up use cases. Here is a little summary::

1. First use case: My keybinding works fine itself. I wrongly thought it didn't work because of the Peek Definition window. If the Peek Definition window was open inside the Search Editor results, then at the widget closure, after the next search task has been completed, the focus was moved from the Search Editor input to its results, like this:

![usecase2](https://github.com/microsoft/vscode/assets/37016009/162f095a-78a6-4669-8791-b4d9eecdfe39)


2. Second use case: If the Search Editor has been opened from the search view ("Open in editor" button) it focus the Search Editor results, like this:

![usecase1](https://github.com/microsoft/vscode/assets/37016009/7d651bde-b817-4a46-97fd-0c8a236a0624)

Both use cases could be resolved by adding the focusResults option to Search Editor options, so we could end up with a consistent behavior no matter where the triggerSearch has been fired from.

andreamah commented:
I don't think that this would be the proper way to control where the focus should go, since this only relates to using the keybinding (which it doesn't trigger when you do things like "open in editor"). This would be better suited to a setting.

chandra1109 commented:
But the code which we had has been already moved to PD .
We cant make changes until QA is done
Regards .

On Thu, 4 Apr 2024 at 2:39 AM, Andrea Mah ***@***.***> wrote:

> I don't think that this would be the proper way to control where the focus
> should go, since this only relates to using the keybinding (which it
> doesn't trigger when you do things like "open in editor"). This would be
> better suited to a setting.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/vscode/pull/205914#issuecomment-2035595130>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZMD43EXYNONZZHDOZ3YMRTY3RVYRAVCNFSM6AAAAABDTZRLOKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMZVGU4TKMJTGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

mkasenberg commented:
> I don't think that this would be the proper way to control where the focus should go, since this only relates to using the keybinding (which it doesn't trigger when you do things like "open in editor"). This would be better suited to a setting.

I would like to cover 3 use cases that I know so far that triggers the `triggerSearch()` function:
1. a key binding,
2. "open in editor",
3. changing the input of the search editor.

I refactored again to make the `search.searchEditor.focusResults` setting an optional one that allows to set the preferred default focusResults value (so it covers each of my cases 1, 2, 3).

I changed back the declaration of
```
async triggerSearch(_options?: { resetCursor?: boolean; delay?: number; focusResults?: boolean })` to the original one to make use of the `args.focusResults
```
just in case someone still would like to have multiple keybindings with different behavior than the default one. (so it covers my case 1).

mkasenberg commented:
> But the code which we had has been already moved to PD . We cant make changes until QA is done Regards .

Sorry, I am not familiar with the vscode contribution workflow yet. What does PD and QA mean? Do I have to do something or it just means that the repo is in freezed state before next release?

andreamah commented on src/vs/workbench/contrib/searchEditor/browser/searchEditor.ts:
```diff
@@ -494,6 +502,8 @@ export class SearchEditor extends AbstractTextCodeEditor<SearchEditorViewState>
 				}
 				if (options.focusResults) {
 					this.searchResultEditor.focus();
+				} else {
+					this.focusSearchInput();
```
Is this necessary? I believe that `search.searchEditor.focusResults: false` should mean that it should just keep focus wherever it already is.

mkasenberg replied:
Yes, this part is a workaround for a use case where a PeekViewWidget was open in previous Search Editor results and after a new search is completed, at the widget closure, the ReferencesController moves focus to results, instead of leaving it at Search Editor input. I will add a comment here too.

andreamah replied:
Sorry, I'm a bit confused- do you want the references view to stay open on refresh? And focusing the search input keeps it open?

mkasenberg replied:
No, no, the references view is being closed on refresh as it should be. I just meant that some callback sequence is fired when closing this view and it moves the focus to the editor results. When I close the references view manually before new search, the focus stays where I want it to be, at the search input.

andreamah replied:
So are you saying that closing the references view is moving the focus to the editor inadvertently? That seems like a bug with the references peek to me. So the sequence without this line is:
1. refresh of search editor closes the peek editor
2. closing the peek editor sets the cursor to where the peek editor was
3. cursor is now where the peek editor was, even if the focus was on the input previously?

Are you sure that this isn't caused by something like the `options.resetCursor`?

mkasenberg replied:
Well summed up.

After further testing with fresh eyes, it looks like this is the line that moves the focus:
https://github.com/microsoft/vscode/blob/fca05df7d4a9ce627919af8cd9fc3a3f2e14e029/src/vs/editor/contrib/gotoSymbol/browser/peek/referencesController.ts#L110

The usage of `options.resetCursor` moves the cursor pointer within the text area, but does not change the focus:
https://github.com/microsoft/vscode/blob/fca05df7d4a9ce627919af8cd9fc3a3f2e14e029/src/vs/workbench/contrib/searchEditor/browser/searchEditor.ts#L492-L495

The definition of the `closeWidget` method has enabled the editor focusing by default:

https://github.com/microsoft/vscode/blob/fca05df7d4a9ce627919af8cd9fc3a3f2e14e029/src/vs/editor/contrib/gotoSymbol/browser/peek/referencesController.ts#L229-L240

so when I test with the line set to
```
this.closeWidget(false); 
```
the focus stays in the input where I want it to be. But I am afraid of this change, because someone could complain that the exit button of the peek viewer doesn't work as before because it no longer returns focus to the editor.

So I would stay with the `this.focusSearchInput();` call because it looks like a safe compromise.

andreamah replied:
How about changing the line that calls it to:
```typescript
controller?.closeWidget(this.searchResultEditor.hasTextFocus());
```
Then, it will act differently the same if the editor has focus. Otherwise, it will not try to focus the editor.

andreamah replied:
I think I'm a bit confused about what you mean, since that line in `referencesController` is called from 
https://github.com/microsoft/vscode/blob/660018a75f944cd9cc27073eef4fc6918ef7ae23/src/vs/workbench/contrib/searchEditor/browser/searchEditor.ts#L615
which already specifies `focusEditor = false`?

andreamah replied:
Oh, I see in the stack trace: `closeWidget` gets called twice, where the second time gets called with `true` regardless of what the original call had. Let me investigate.

andreamah commented on src/vs/workbench/contrib/search/browser/search.contribution.ts:
```diff
@@ -328,6 +328,11 @@ configurationRegistry.registerConfiguration({
+		'search.searchEditor.focusResults': {
+			type: 'boolean',
+			default: false,
+			markdownDescription: nls.localize('search.searchEditor.focusResults', "Focus the searchEditor results instead of the searchEditor input.")
```
since this user-facing, we can use "Search Editor" instead of "searchEditor"

andreamah commented on src/vs/workbench/contrib/searchEditor/browser/searchEditor.ts:
```diff
@@ -482,6 +482,14 @@ export class SearchEditor extends AbstractTextCodeEditor<SearchEditorViewState>
 	async triggerSearch(_options?: { resetCursor?: boolean; delay?: number; focusResults?: boolean }) {
+		const focusResults = this.searchConfig.searchEditor.focusResults;
+
+		if (_options === undefined) {
+			_options = { focusResults: focusResults };
+		} else if (_options.focusResults === undefined) {
+			_options.focusResults = focusResults;
+		}
```
This logic is saying "if options don't define `focusResult` field, then use the setting", right? Maybe a comment would make this clearer.

andreamah reviewed (changes requested):
This is really shaping up- just a few comments!

andreamah commented:
> > But the code which we had has been already moved to PD . We cant make changes until QA is done Regards .
> 
> Sorry, I am not familiar with the vscode contribution workflow yet. What does PD and QA mean? Do I have to do something or it just means that the repo is in freezed state before next release?

I also do not know what the original comment is discussing. There is a period right before release where we perform QA and cannot merge PRs, but we can merge any other time to get changes into the next stable release.

andreamah commented:
> Sorry, I mixed up use cases. Here is a little summary::
> 
> 1. First use case: My keybinding works fine itself. I wrongly thought it didn't work because of the Peek Definition window. If the Peek Definition window was open inside the Search Editor results, then at the widget closure, after the next search task has been completed, the focus was moved from the Search Editor input to its results, like this:
> 
> ![usecase2](https://private-user-images.githubusercontent.com/37016009/318202730-162f095a-78a6-4669-8791-b4d9eecdfe39.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NDA4MzcsIm5iZiI6MTcxNzU0MDUzNywicGF0aCI6Ii8zNzAxNjAwOS8zMTgyMDI3MzAtMTYyZjA5NWEtNzhhNi00NjY5LTg3OTEtYjRkOWVlY2RmZTM5LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA0VDIyMzUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzMTNlNTI2MjdhYTY3ODE5NWRjYzk1NjQwNmM1YzE0MDMxY2VjOGJmNDUxMmZmNTRkYjBjZmY1MGMwNmZiMDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tSF0PHEaNLc9cw_7GjYMR7j5Ik7igCHD23GHv4ihDs4) [ ![usecase2](https://private-user-images.githubusercontent.com/37016009/318202730-162f095a-78a6-4669-8791-b4d9eecdfe39.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NDA4MzcsIm5iZiI6MTcxNzU0MDUzNywicGF0aCI6Ii8zNzAxNjAwOS8zMTgyMDI3MzAtMTYyZjA5NWEtNzhhNi00NjY5LTg3OTEtYjRkOWVlY2RmZTM5LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA0VDIyMzUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzMTNlNTI2MjdhYTY3ODE5NWRjYzk1NjQwNmM1YzE0MDMxY2VjOGJmNDUxMmZmNTRkYjBjZmY1MGMwNmZiMDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tSF0PHEaNLc9cw_7GjYMR7j5Ik7igCHD23GHv4ihDs4) ](https://private-user-images.githubusercontent.com/37016009/318202730-162f095a-78a6-4669-8791-b4d9eecdfe39.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NDA4MzcsIm5iZiI6MTcxNzU0MDUzNywicGF0aCI6Ii8zNzAxNjAwOS8zMTgyMDI3MzAtMTYyZjA5NWEtNzhhNi00NjY5LTg3OTEtYjRkOWVlY2RmZTM5LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA0VDIyMzUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzMTNlNTI2MjdhYTY3ODE5NWRjYzk1NjQwNmM1YzE0MDMxY2VjOGJmNDUxMmZmNTRkYjBjZmY1MGMwNmZiMDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tSF0PHEaNLc9cw_7GjYMR7j5Ik7igCHD23GHv4ihDs4) [ ](https://private-user-images.githubusercontent.com/37016009/318202730-162f095a-78a6-4669-8791-b4d9eecdfe39.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NDA4MzcsIm5iZiI6MTcxNzU0MDUzNywicGF0aCI6Ii8zNzAxNjAwOS8zMTgyMDI3MzAtMTYyZjA5NWEtNzhhNi00NjY5LTg3OTEtYjRkOWVlY2RmZTM5LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA0VDIyMzUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzMTNlNTI2MjdhYTY3ODE5NWRjYzk1NjQwNmM1YzE0MDMxY2VjOGJmNDUxMmZmNTRkYjBjZmY1MGMwNmZiMDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tSF0PHEaNLc9cw_7GjYMR7j5Ik7igCHD23GHv4ihDs4)
> 
> 2. Second use case: If the Search Editor has been opened from the search view ("Open in editor" button) it focus the Search Editor results, like this:
> 
> ![usecase1](https://private-user-images.githubusercontent.com/37016009/318200912-7d651bde-b817-4a46-97fd-0c8a236a0624.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NDA4MzcsIm5iZiI6MTcxNzU0MDUzNywicGF0aCI6Ii8zNzAxNjAwOS8zMTgyMDA5MTItN2Q2NTFiZGUtYjgxNy00YTQ2LTk3ZmQtMGM4YTIzNmEwNjI0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA0VDIyMzUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjYTllZmRjNTIyMTZmMTUyMTNhMjBjNTRmMTRhZjg1OWFhN2E3MTUxZTY1NjAzZDFjNWQzOWFjMzYxZGU4M2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kbnS4R1L_zDx7zLk_XsN3QwNbBRYhYwAX9Woj7ZBG58) [ ![usecase1](https://private-user-images.githubusercontent.com/37016009/318200912-7d651bde-b817-4a46-97fd-0c8a236a0624.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NDA4MzcsIm5iZiI6MTcxNzU0MDUzNywicGF0aCI6Ii8zNzAxNjAwOS8zMTgyMDA5MTItN2Q2NTFiZGUtYjgxNy00YTQ2LTk3ZmQtMGM4YTIzNmEwNjI0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA0VDIyMzUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjYTllZmRjNTIyMTZmMTUyMTNhMjBjNTRmMTRhZjg1OWFhN2E3MTUxZTY1NjAzZDFjNWQzOWFjMzYxZGU4M2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kbnS4R1L_zDx7zLk_XsN3QwNbBRYhYwAX9Woj7ZBG58) ](https://private-user-images.githubusercontent.com/37016009/318200912-7d651bde-b817-4a46-97fd-0c8a236a0624.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NDA4MzcsIm5iZiI6MTcxNzU0MDUzNywicGF0aCI6Ii8zNzAxNjAwOS8zMTgyMDA5MTItN2Q2NTFiZGUtYjgxNy00YTQ2LTk3ZmQtMGM4YTIzNmEwNjI0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA0VDIyMzUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjYTllZmRjNTIyMTZmMTUyMTNhMjBjNTRmMTRhZjg1OWFhN2E3MTUxZTY1NjAzZDFjNWQzOWFjMzYxZGU4M2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kbnS4R1L_zDx7zLk_XsN3QwNbBRYhYwAX9Woj7ZBG58) [ ](https://private-user-images.githubusercontent.com/37016009/318200912-7d651bde-b817-4a46-97fd-0c8a236a0624.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NDA4MzcsIm5iZiI6MTcxNzU0MDUzNywicGF0aCI6Ii8zNzAxNjAwOS8zMTgyMDA5MTItN2Q2NTFiZGUtYjgxNy00YTQ2LTk3ZmQtMGM4YTIzNmEwNjI0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA0VDIyMzUzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjYTllZmRjNTIyMTZmMTUyMTNhMjBjNTRmMTRhZjg1OWFhN2E3MTUxZTY1NjAzZDFjNWQzOWFjMzYxZGU4M2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kbnS4R1L_zDx7zLk_XsN3QwNbBRYhYwAX9Woj7ZBG58)
> 
> Both use cases could be resolved by adding the focusResults option to Search Editor options, so we could end up with a consistent behavior no matter where the triggerSearch has been fired from.

@mkasenberg would you mind creating two different issues for these and linking them to this PR? Just for verification purposes, so that we have an issue for each thing that gets fixed :)

andreamah commented:
For the first problem that you described, I've created https://github.com/microsoft/vscode/issues/214285, and am waiting on a fix (https://github.com/microsoft/vscode/pull/214289) so that you don't need to add extra code to fix this. If you create an issue, you can probably just create one for the second issue.

andreamah commented on src/vs/workbench/contrib/search/browser/search.contribution.ts:
```diff
@@ -329,6 +329,11 @@ configurationRegistry.registerConfiguration({
 			default: 1,
 			markdownDescription: nls.localize('search.searchEditor.defaultNumberOfContextLines', "The default number of surrounding context lines to use when creating new Search Editors. If using `#search.searchEditor.reusePriorSearchConfiguration#`, this can be set to `null` (empty) to use the prior Search Editor's configuration.")
 		},
+		'search.searchEditor.focusResults': {
+			type: 'boolean',
+			default: false,
+			markdownDescription: nls.localize('search.searchEditor.focusResults', "Focus the Search Editor results instead of the Search Editor input.")
+		},
```
I got https://github.com/microsoft/vscode/pull/214289 merged in, so this PR should be pretty much ready to be merged in. One final thing: maybe we should make this clearer that this is where it focuses on search. So maybe calling it `search.searchEditor.focusResultsOnSearch` and mentioning that the setting applies when you trigger a search.

andreamah reviewed (approved):
Awesome, thanks for your patience! And good catch with the widget closing bug!
